python/ray/experimental/dag/py_obj_scanner.py

```diff
@@ -0,0 +1,69 @@
+import ray
+
+import copyreg
+import uuid
+import io
+import pickle
+from typing import List, Dict, Any, TypeVar, TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from ray.experimental.dag.dag_node import DAGNode
+
+T = TypeVar("T")
+
+
+class _PyObjScanner(ray.cloudpickle.CloudPickler):
+    """Utility to find and replace DAGNodes in Python objects.
+
+    This uses pickle to walk the PyObj graph and find first-level DAGNode
+    instances on ``find_nodes()``. The caller can then compute a replacement
+    table and then replace the nodes via ``replace_nodes()``.
+    """
+
+    # Used in deserialization hooks to reference scanner instances.
+    _instances: Dict[str, "_PyObjScanner"] = {}
+
+    def __init__(self):
+        # Buffer to keep intermediate serialized state.
+        self._buf = io.BytesIO()
+        # List of top-level DAGNodes found during the serialization pass.
+        self._found = None
+        # Replacement table to consult during deserialization.
+        self._replace_table: Dict["DAGNode", T] = None
+        # UUID of this scanner.
+        self._uuid = uuid.uuid4().hex
+        _PyObjScanner._instances[self._uuid] = self
+        # Register pickler override for DAGNode types.
+        from ray.experimental.dag.task_node import TaskNode
+        from ray.experimental.dag.actor_node import ActorNode, ActorMethodNode
+        self.dispatch_table[TaskNode] = self._reduce_dag_node
+        self.dispatch_table[ActorNode] = self._reduce_dag_node
+        self.dispatch_table[ActorMethodNode] = self._reduce_dag_node
+        super().__init__(self._buf)
+
+    def find_nodes(self, obj: Any) -> List["DAGNode"]:
+        """Find top-level DAGNodes."""
+        assert self._found is None, "find_nodes cannot be called twice"
+        self._found = []
+        self.dump(obj)
+        return self._found
+
+    def replace_nodes(self, table: Dict["DAGNode", T]) -> Any:
+        """Replace previously found DAGNodes per the given table."""
+        assert self._found is not None, "find_nodes must be called first"
+        self._replace_table = table
+        self._buf.seek(0)
+        return pickle.load(self._buf)
+
+    def _replace_index(self, i: int) -> "DAGNode":
+        return self._replace_table[self._found[i]]
+
+    def _reduce_dag_node(self, obj):
+        uuid = self._uuid
+        index = len(self._found)
+        res = (lambda i: _PyObjScanner._instances[uuid]._replace_index(i)), (index,)
```

Comment:
we should not reach into private fields of `_PyObjScanner`, if this is necessary make it a public interface.

Reply:
This is a private method call being generated from a private method, so it's fine.

Reply:
Oh, I didn't realize this is from the same class...